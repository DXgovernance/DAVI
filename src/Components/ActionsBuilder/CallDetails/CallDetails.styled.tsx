import { MetadataTag } from '../SupportedActions/common/infoLine';
import { DetailRow } from '../SupportedActions/common/summary';
import styled from 'styled-components';
import { Box } from 'old-components/common/Box';

export const ParamTag = styled(MetadataTag)`
  border-radius: ${({ theme }) => theme.radii.pill};
  margin: 0 0.25rem;
  color: ${({ color }) => color};
`;

export const ParamTitleRow = styled(DetailRow)`
  margin-bottom: 0.75rem;
`;

export const ParamTitleTag = styled(MetadataTag)`
  display: inline-block;
  padding: 0.375rem 0.5rem;
  color: ${({ color }) => color};
`;

export const ActionParamRow = styled(Box)`
  margin-bottom: 1.5rem;
`;

export const ParamDetail = styled(Box)`
  color: ${({ theme }) => theme.colors.proposalText.grey};
  overflow-wrap: break-word;
`;
