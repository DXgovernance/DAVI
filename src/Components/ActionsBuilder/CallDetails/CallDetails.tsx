import { ActionViewProps } from '../SupportedActions';
import { BigNumber } from 'ethers';
import { Button } from 'old-components/Guilds/common/Button';
import { Box } from 'Components/Primitives/Layout';
import UnstyledLink from 'Components/Primitives/Links/UnstyledLink';
import { FiExternalLink } from 'react-icons/fi';
import { useTheme } from 'styled-components';
import { Divider } from 'old-components/Guilds/common/Divider';
import {
  ActionParamRow,
  ParamDetail,
  ParamTag,
  ParamTitleRow,
  ParamTitleTag,
} from './CallDetails.styled';
import { useTranslation } from 'react-i18next';

export const CallDetails: React.FC<ActionViewProps> = ({
  decodedCall,
  approveSpendTokens,
}) => {
  const theme = useTheme();
  const { t } = useTranslation();

  function renderByParamType(type: string, value: any) {
    if (!type || !value) return null;

    if (type === 'address') {
      return (
        <UnstyledLink to="#">
          <ParamDetail>
            {value} <FiExternalLink size={16} />
          </ParamDetail>
        </UnstyledLink>
      );
    }

    if (type.startsWith('uint') || type.startsWith('int')) {
      if (Array.isArray(value)) {
        let valuesStringFromArray: string = '[';

        value.forEach((element, index) => {
          let separator = ', ';
          if (index === 0) separator = '';

          const elementBigNumber = BigNumber.from(element).toString();
          return (valuesStringFromArray += separator += elementBigNumber);
        });
        valuesStringFromArray += ']';

        return <ParamDetail>{valuesStringFromArray}</ParamDetail>;
      }

      return <ParamDetail>{BigNumber.from(value).toString()}</ParamDetail>;
    }

    return <ParamDetail>{value}</ParamDetail>;
  }

  return (
    <>
      {!!approveSpendTokens && (
        <>
          <ActionParamRow>
            <Box>
              approve ({' '}
              <ParamTag color={theme?.colors?.params?.[0]}>address</ParamTag>
              {', '}
              <ParamTag color={theme?.colors?.params?.[1]}>uint256</ParamTag> )
            </Box>
          </ActionParamRow>
          <ActionParamRow>
            <ParamTitleRow>
              <ParamTitleTag color={theme?.colors?.params?.[0]}>
                spender <em>(address)</em>
              </ParamTitleTag>
            </ParamTitleRow>
            {renderByParamType('address', decodedCall.from)}
          </ActionParamRow>
          <ActionParamRow>
            <ParamTitleRow>
              <ParamTitleTag color={theme?.colors?.params?.[1]}>
                amount <em>(uint256)</em>
              </ParamTitleTag>
            </ParamTitleRow>
            {renderByParamType('uint256', approveSpendTokens?.amount)}
          </ActionParamRow>
          <Divider style={{ marginBottom: '2rem' }} />
        </>
      )}
      <ActionParamRow>
        <Box>
          {decodedCall?.function?.name} (
          {decodedCall?.function?.inputs.map((param, index, params) => (
            <span key={index}>
              <ParamTag color={theme?.colors?.params?.[index]}>
                {param?.type}
              </ParamTag>
              {index < params.length - 1 && <span> , </span>}
            </span>
          ))}
          )
        </Box>
      </ActionParamRow>

      {decodedCall?.function?.inputs?.map((param, index) => (
        <ActionParamRow key={index}>
          <ParamTitleRow>
            <ParamTitleTag color={theme?.colors?.params?.[index]}>
              {param.name} <em>({param.type})</em>
            </ParamTitleTag>
            {param.type === 'bytes' && (
              <Button variant="secondary">{t('decode')}</Button>
            )}
          </ParamTitleRow>

          {renderByParamType(param.type, decodedCall?.args?.[param.name])}
        </ActionParamRow>
      ))}
    </>
  );
};

export default CallDetails;
