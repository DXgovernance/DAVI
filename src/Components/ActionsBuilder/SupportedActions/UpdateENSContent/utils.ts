import contenthash from 'content-hash';
import { MAINNET_ID, LOCALHOST_ID, GOERLI_ID } from 'utils';
import { utils } from 'ethers';

const DEFAULT_NAMEHASH =
  '0x0000000000000000000000000000000000000000000000000000000000000000';

export const convertToNameHash = (name: string) => {
  if (!name) {
    return DEFAULT_NAMEHASH;
  }
  return utils.namehash(name);
};

export const convertToContentHash = (ipfsHash: string) => {
  if (!ipfsHash) {
    return '';
  }
  return contenthash.fromIpfs(ipfsHash);
};

export const convertToIpfsHash = (contentHash: string) => {
  if (!contentHash) return null;
  return contenthash.decode(contentHash);
};

export const getIpfsUrl = (ipfsHash: string) => {
  const ipfsRoot = 'ipfs://';
  return `${ipfsRoot}${ipfsHash}`;
};

export const isValidChainId = (chainId: number) => {
  if (!chainId || chainId === LOCALHOST_ID) {
    return MAINNET_ID;
  } else {
    return chainId;
  }
};

export const isAvailableOnENS = (chainId: number) => {
  const validChainId = isValidChainId(chainId);
  const ensNetworks = [MAINNET_ID, GOERLI_ID];
  return ensNetworks.includes(validChainId);
};
