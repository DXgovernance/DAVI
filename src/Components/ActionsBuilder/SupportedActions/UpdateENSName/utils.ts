import namehash from 'eth-ens-namehash';
import contenthash from 'content-hash';
import { useState, useEffect } from 'react';

export const convertToNameHash = (name: string) => namehash.hash(name);

export const convertToContentHash = (ipfsHash: string) =>
  contenthash.fromIpfs(ipfsHash);

export const convertToIPFSHash = (contentHash: string) => {
  if (!contentHash) return null;
  return contenthash.decode(contentHash);
};

export const useDebounce = (value: string, delay: number) => {
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value, delay]);

  return debouncedValue;
};
