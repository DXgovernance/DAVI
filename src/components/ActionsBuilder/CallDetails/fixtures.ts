import {
  ApproveSendTokens,
  DecodedCall,
  SupportedAction,
} from 'components/ActionsBuilder/types';
import ERC20 from 'contracts/ERC20.json';
import { BigNumber, utils } from 'ethers';

const ERC20Contract = new utils.Interface(ERC20.abi);

export const emptyDecodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '',
  value: BigNumber.from(0),
  args: {
    _to: '',
    _value: BigNumber.from(0),
  },
};

export const decodedCallMock: DecodedCall = {
  from: '0x0000000000000000000000000000000000000000',
  callType: SupportedAction.ERC20_TRANSFER,
  function: ERC20Contract.getFunction('transfer'),
  to: '0x95a223299319022a842d0dfe4851c145a2f615b9',
  value: BigNumber.from(0),
  args: {
    _to: '0x0000000000000000000000000000000000000000',
    _value: BigNumber.from(1000),
  },
};

export const approvetokenSpendingMock: ApproveSendTokens = {
  callType: SupportedAction.GENERIC_CALL,
  from: '0x0000000000000000000000000000000000000000',
  to: '0x0000000000000000000000000000000000000000',
  value: BigNumber.from(0),
  function: null,
  args: {},
  amount: BigNumber.from(1000),
  token: '0x3f943f38b2fbe1ee5daf0516cecfe4e0f8734351',
};
