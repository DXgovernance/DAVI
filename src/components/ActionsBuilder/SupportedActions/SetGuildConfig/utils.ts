import { FIELDS } from './constants';
import { BigNumber } from 'ethers';
import { SetGuildConfigFields, ControlField } from './types';

export const bn = (n: string | number | BigNumber) => BigNumber.from(n);

type UpdatedValuesReturn = {
  [key in ControlField]: {
    newValue: BigNumber;
    currentValue: BigNumber;
  };
};
export const getUpdatedValues = (
  current: SetGuildConfigFields,
  modifyed: SetGuildConfigFields
): UpdatedValuesReturn | {} => {
  if (!current || !modifyed) return {};
  return Object.keys(current).reduce((acc, key) => {
    if (!FIELDS.map(f => f.name).includes(key)) return acc;
    const currentValue = bn(current[key] || 0);
    const newValue = bn(modifyed[key] || 0);

    if (!currentValue.eq(newValue)) {
      return {
        ...acc,
        [key]: { newValue, currentValue },
      };
    }
    return acc;
  }, {});
};

export const pickUpdatedOrDefaultValue = (
  current: BigNumber,
  modifyed: BigNumber
): BigNumber => {
  if (BigNumber.isBigNumber(current)) {
    if (BigNumber.isBigNumber(modifyed) && !bn(modifyed).eq(current)) {
      return bn(modifyed);
    }
    return bn(current);
  } else if (BigNumber.isBigNumber(modifyed)) {
    return bn(modifyed);
  }
  return bn(0);
};
