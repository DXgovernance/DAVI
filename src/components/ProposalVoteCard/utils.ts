import { ConfirmVoteProposalProps, VoteOnProposalProps } from './types';

export const voteOnProposal = ({
  votingPowerAtProposalSnapshot,
  votingPowerAtProposalCurrentSnapshot,
}: VoteOnProposalProps) => {
  const hasNoVotingPower =
    votingPowerAtProposalSnapshot &&
    Number(votingPowerAtProposalSnapshot?.toString()) <= 0;

  const hasVotingPowerAtCurrentSnapshot =
    votingPowerAtProposalCurrentSnapshot &&
    Number(votingPowerAtProposalCurrentSnapshot?.toString()) > 0;
  return {
    hasNoVotingPower,
    hasVotingPowerAtCurrentSnapshot,
  };
};

export const confirmVoteProposal = ({
  proposal,
  contract,
  selectedOption,
  userVotingPower,
  createTransaction,
  cb,
}: ConfirmVoteProposalProps) => {
  createTransaction(
    `Vote on proposal ${proposal?.title}`,
    async () => contract.setVote(proposal?.id, selectedOption, userVotingPower),
    true,
    cb
  );
};

export const getOptionLabel = ({ metadata, optionKey, t }) => {
  const metadataLabel = metadata?.voteOptions?.[optionKey];
  return metadataLabel
    ? metadataLabel
    : Number(optionKey) === 0
    ? t('against', { defaultValue: 'Against' })
    : t('option', { optionKey });
};
