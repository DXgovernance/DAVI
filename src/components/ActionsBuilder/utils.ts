import { DecodedAction } from './types';
import { Permission } from './types';
import Web3 from 'web3';
import { NULL_SIGNATURE } from 'utils';
import { SupportedAction } from './types';

const web3 = new Web3();

export const getPermissionArgs = (
  decodedActions: DecodedAction[]
): Permission[] => {
  const permissions = decodedActions.map(decodedAction => {
    // Native transfer of ETH does not have a function
    if (
      decodedAction?.decodedCall?.callType ===
        SupportedAction.NATIVE_TRANSFER ||
      decodedAction?.decodedCall?.callType === SupportedAction.RAW_TRANSACTION
    ) {
      return {
        from: decodedAction?.decodedCall?.from,
        to: decodedAction?.decodedCall?.to,
        callType: decodedAction?.decodedCall?.callType,
        functionSignature: NULL_SIGNATURE,
      };
    }
    const functionInputs = decodedAction?.decodedCall?.function?.inputs?.map(
      input => input.type
    );
    const functionName = decodedAction?.decodedCall?.function?.name;
    const functionSignature = web3.eth.abi.encodeFunctionSignature(
      `${functionName}(${functionInputs.join(',')})`
    );
    return {
      from: decodedAction?.decodedCall?.from,
      to: decodedAction?.decodedCall?.to,
      callType: decodedAction?.decodedCall?.callType,
      functionSignature,
    };
  });
  return permissions;
};
