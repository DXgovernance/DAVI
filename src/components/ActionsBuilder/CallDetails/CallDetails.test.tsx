import { BigNumber } from 'ethers';
import { render } from 'utils/tests';
import { CallDetails } from './CallDetails';
import {
  approvetokenSpendingMock,
  decodedCallMock,
  emptyDecodedCallMock,
} from './fixtures';

const mockBigNumber = BigNumber.from(0);

jest.mock('hooks/Guilds/ens/useENSAvatar', () => ({
  __esModule: true,
  default: () => ({
    avatarUri: 'test',
    imageUrl: 'test',
    ensName: 'test.eth',
  }),
}));

jest.mock('utils', () => ({
  getNetworkById: () => ({
    nativeAsset: {
      symbol: 'TST',
    },
  }),
  preventEmptyString: () => mockBigNumber,
}));

jest.mock('wagmi', () => ({
  useNetwork: () => ({ chain: { id: 1 } }),
  chain: {
    mainnet: {},
  },
}));

describe('CallDetails', () => {
  it('Should match', () => {
    const { container } = render(<CallDetails decodedCall={decodedCallMock} />);
    expect(container).toMatchSnapshot();
  });

  it('Should match with empty data', () => {
    const { container } = render(
      <CallDetails
        decodedCall={emptyDecodedCallMock}
        approveSpendTokens={approvetokenSpendingMock}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
