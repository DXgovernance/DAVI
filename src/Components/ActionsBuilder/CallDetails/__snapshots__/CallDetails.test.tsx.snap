// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallDetails Should match 1`] = `
<div>
   
  <div
    class="sc-dQoVA sc-fodVek ijKhIZ jPSkGh"
  >
    <div
      class="sc-hKgJUU dpFBPj"
    >
      transfer
       (
      <span>
        <span
          class="sc-jrAFXE sc-bqyKOL hTcuJV eHjsis"
          color="#9575CD"
        >
          address
        </span>
        <span>
           , 
        </span>
      </span>
      <span>
        <span
          class="sc-jrAFXE sc-bqyKOL hTcuJV bHUmRM"
          color="#F06292"
        >
          uint256
        </span>
      </span>
      )
    </div>
  </div>
  <div
    class="sc-dQoVA sc-fodVek ijKhIZ jPSkGh"
  >
    <div
      class="sc-hKgJUU sc-kEjbQP sc-kstqJO dpFBPj fUIUMn hZygGC"
    >
      <span
        class="sc-jrAFXE sc-hBEYId hTcuJV gvGmHF"
        color="#9575CD"
      >
        recipient
         
        <em>
          (
          address
          )
        </em>
      </span>
    </div>
  </div>
  <div
    class="sc-dQoVA sc-fodVek ijKhIZ jPSkGh"
  >
    <div
      class="sc-hKgJUU sc-kEjbQP sc-kstqJO dpFBPj fUIUMn hZygGC"
    >
      <span
        class="sc-jrAFXE sc-hBEYId hTcuJV iwJLYR"
        color="#F06292"
      >
        amount
         
        <em>
          (
          uint256
          )
        </em>
      </span>
    </div>
  </div>
   
</div>
`;

exports[`CallDetails Should match with empty data 1`] = `
<div>
   
  <div
    class="sc-dQoVA sc-fodVek ijKhIZ jPSkGh"
  >
    <div
      class="sc-hKgJUU dpFBPj"
    >
      approve (
       
      <span
        class="sc-jrAFXE sc-bqyKOL hTcuJV eHjsis"
        color="#9575CD"
      >
        address
      </span>
      , 
      <span
        class="sc-jrAFXE sc-bqyKOL hTcuJV bHUmRM"
        color="#F06292"
      >
        uint256
      </span>
       )
    </div>
  </div>
  <div
    class="sc-dQoVA sc-fodVek ijKhIZ jPSkGh"
  >
    <div
      class="sc-hKgJUU sc-kEjbQP sc-kstqJO dpFBPj fUIUMn hZygGC"
    >
      <span
        class="sc-jrAFXE sc-hBEYId hTcuJV gvGmHF"
        color="#9575CD"
      >
        spender 
        <em>
          (address)
        </em>
      </span>
    </div>
    <a
      class="sc-iBPTik dFVotB"
      href="/"
    >
      <div
        class="sc-dQoVA sc-fFucqa ijKhIZ gzJfIk"
      >
        0x0000000000000000000000000000000000000000
         
        <svg
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
          />
          <polyline
            points="15 3 21 3 21 9"
          />
          <line
            x1="10"
            x2="21"
            y1="14"
            y2="3"
          />
        </svg>
      </div>
    </a>
  </div>
  <div
    class="sc-dQoVA sc-fodVek ijKhIZ jPSkGh"
  >
    <div
      class="sc-hKgJUU sc-kEjbQP sc-kstqJO dpFBPj fUIUMn hZygGC"
    >
      <span
        class="sc-jrAFXE sc-hBEYId hTcuJV iwJLYR"
        color="#F06292"
      >
        amount 
        <em>
          (uint256)
        </em>
      </span>
    </div>
    <div
      class="sc-dQoVA sc-fFucqa ijKhIZ gzJfIk"
    >
      1000
    </div>
  </div>
  <div
    class="sc-fubCzh dIZEAK"
    style="margin-bottom: 2rem;"
  />
  <div
    class="sc-dQoVA sc-fodVek ijKhIZ jPSkGh"
  >
    <div
      class="sc-hKgJUU dpFBPj"
    >
      transfer
       (
      <span>
        <span
          class="sc-jrAFXE sc-bqyKOL hTcuJV eHjsis"
          color="#9575CD"
        >
          address
        </span>
        <span>
           , 
        </span>
      </span>
      <span>
        <span
          class="sc-jrAFXE sc-bqyKOL hTcuJV bHUmRM"
          color="#F06292"
        >
          uint256
        </span>
      </span>
      )
    </div>
  </div>
  <div
    class="sc-dQoVA sc-fodVek ijKhIZ jPSkGh"
  >
    <div
      class="sc-hKgJUU sc-kEjbQP sc-kstqJO dpFBPj fUIUMn hZygGC"
    >
      <span
        class="sc-jrAFXE sc-hBEYId hTcuJV gvGmHF"
        color="#9575CD"
      >
        recipient
         
        <em>
          (
          address
          )
        </em>
      </span>
    </div>
  </div>
  <div
    class="sc-dQoVA sc-fodVek ijKhIZ jPSkGh"
  >
    <div
      class="sc-hKgJUU sc-kEjbQP sc-kstqJO dpFBPj fUIUMn hZygGC"
    >
      <span
        class="sc-jrAFXE sc-hBEYId hTcuJV iwJLYR"
        color="#F06292"
      >
        amount
         
        <em>
          (
          uint256
          )
        </em>
      </span>
    </div>
  </div>
   
</div>
`;
